### PR TITLE
Run fetches on separate thread when part of merge/mutate

### LINKED ISSUE #44384: Fetches executing on merge pool are starving merges leading to continuously increasing replication delay
When a node falls back behind its peer replicas, it tries to catch up using the information in its replication queue.

It can do so using two main strategies: merging parts locally to reproduce merges that already happened on other replicas, or fetching the already merged parts directly from other replicas.

The strategy is chosen using various heuristics as well as the the following settings:

`prefer_fetch_merged_part_time_threshold`, `prefer_fetch_merged_part_size_threshold`.

While typically fetches and merges happen on different thread pools, there seems to be a condition https://github.com/ClickHouse/ClickHouse/blame/6f7bba159f003ec0ed25c4083b92aee89e9c795b/src/Storages/MergeTree/MergeFromLogEntryTask.cpp#L131 which makes a merge turn into a fetch.

Often, it occurs after the replica reboots after falling significantly behind.

The problem/bug since #25165 is that merges can yield to other merges while the fetches cannot.

When long fetches start executing on the merge pool, merges are getting delayed/starved. Some merges are getting suspended for very long time, other are being delayed and turned into fetches which in turn starve merges...

### PR DESCRIPTION
**Pull request by nvartolomei**
As described in https://github.com/ClickHouse/ClickHouse/issues/44384,
fetches executed on the merge/mutate pool may starve higher priority
tasks. Here we schedule the actual fetch part on separate thread and
yield the control. Periodically, we'll check if the fetch is finished.

This fix is meant as a quick fix for the starving/yielding problem due
to fetch being non-cooperative.

We avoid scheduling the task on the fetch pool as that pool is usually
usually much smaller which causes some changes in the behaviour of
mutations. E.g. we may fail to schedule the mutation immediately and
we have to either retry in a loop (not too nice) or return a retriable
error (we break tests that expect mutations to succeed on first try).

Fixes https://github.com/ClickHouse/ClickHouse/issues/44384.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Always schedule fetches on the fetch pool

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by nvartolomei** [comment]
@nikitamikhaylov wdyt? :)

**Comment by nvartolomei** [comment]
This does not work as priority of all fetch tasks is implicitly 0 (highest priority) and they don't yield to merge tasks anyway. We can't even compute the right priority as we might not have all the parts locally and we don't know the size without it. We could change the replication log protocol, or update the priority after `fetchPart` gets the metadata but after experimenting a bit with it I still don't think that wouldn't improve the situation described in #44384 by much. Continuing to investigate.

File: src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp [hunk 1]
```diff
@@ -146,10 +148,26 @@ bool ReplicatedMergeMutateTaskBase::executeImpl()
 
     auto execute_fetch = [&] (bool need_to_check_missing_part) -> bool
     {
-        if (storage.executeFetch(entry, need_to_check_missing_part))
-            return remove_processed_entry();
+        /// Schedule fetch on own thread so that we could yield from the current
+        /// merge task and execute other higher priority merges if any.
+        fetch_thread = std::make_unique<ThreadFromGlobalPool>([this, need_to_check_missing_part]()
+        {
+            try
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp:155]: Need to add the thread name and other per-thread stuff.

**Comment by alexey-milovidov** [comment]
Don't we already limit the number of concurrent fetches, e.g., max 8 fetches and 32 merges?

**Pull request closed without merging**